### PR TITLE
Add possibility to attempt retry only if a given condition is true

### PR DESCRIPTION
**Pull request by df5602**
Hi,

That's a very nice looking and useful crate 👍.

For my use case however I would like to retry only if certain conditions are true. (See [https://github.com/df5602/bingers/blob/3c132fb4a04c0ecd9bba113caa861dd0e077fdd8/src/tvmaze_api.rs]() lines 176ff for an example.)

So I've added the possibility to give a predicate to the Retry future (similar to `filter()` with iterators). Retry is only attempted if the predicate returns true. I thought, I'd open a pull request in case you find this a useful feature to have.

Unfortunately I've had to box the closure (otherwise it's not possible to instantiate `Retry` if no condition is given). Maybe there's a better way to achieve this?

Feel free to suggest a better API and other improvements if you have any better ideas.

### CODE REVIEW

**Comment by srijs** [comment]
Hi, and thanks for your PR!

Considering the API, what do you think about solving the boxing issue by introducing a separate struct `RetryIf<I, A, F>` for your use-case?

Maybe we can figure out if there's a way to re-use the delay/retry logic between the two structs `Retry` and `RetryIf`, I think that would be ideal. For instance, `RetryIf` could wrap `Retry` and add the additional logic on top.

**Comment by df5602** [comment]
Hi,

I've tried that approach. I ran into problems with re-using the logic though: Initially I thought I could only implement `poll()` for the new struct and in there call the methods of the original struct. However `retry()` and `attempt()` internally call `self.poll()` which is the wrong `poll()` at this point.
In C++ I'd use inheritance to solve this problem, but I don't know how you would approach this in Rust.

However I've had an idea for a different approach. See [this branch](https://github.com/df5602/rust-tokio-retry/commit/22a06e171d3d766523b9369c18b6938ea44f41ca) for a working implementation.

Basically I created a new `Condition<E>` trait and implemented it for `FnMut(&E) -> bool`s (inspired by your `Action` trait). I then moved the logic from the `Retry<I, A>` struct to a new `RetryIf<I, A, C>` struct. `Retry<I, A>` now contains a `RetryIf<I, A, RetryAlways>` as a field and delegates the calls to `poll()` to `RetryIf`.

What do you think of this approach?

**Comment by srijs** [comment]
I guess `RetryIf` wrapping retry could work if you also create a wrapper for `Action`, but I like your approach as well!

As a small nitpick, I don't think you need the `RetryAlways` struct, `Retry` could just contain `RetryIf<I, A, fn(A::Error) -> bool>`. Might be worth a try.

Feel free to update the PR with the changes you made on the branch, then we can wrap up the last details!

**Comment by df5602** [comment]
I merged the branch into the PR.

> As a small nitpick, I don't think you need the RetryAlways struct, Retry could just contain RetryIf<I, A, fn(A::Error) -> bool>. Might be worth a try.

The result looks kind of strange. I'm not sure whether I took a wrong turn somewhere, but this is what I had to do to make it compile. Is this what you had in mind?

Regarding the API: I originally named the constructor `with_condition()` because it was part of the `Retry` struct. Now that it has it's own struct, I think it would make sense to rename it to `spawn()` for similarity between the two structs. What do you think?

File: src/future.rs [hunk 1]
```diff
@@ -75,26 +76,57 @@ enum RetryFuturePoll<A> where A: Action {
...
 /// Future that drives multiple attempts at an action via a retry strategy.
 pub struct Retry<I, A> where I: Iterator<Item=Duration>, A: Action {
+    retry_if: RetryIf<I, A, fn(&A::Error) -> bool>
+}
+
+impl<I, A> Retry<I, A> where I: Iterator<Item=Duration>, A: Action {
+    pub fn spawn<T: IntoIterator<IntoIter=I, Item=Duration>>(handle: Handle, strategy: T, action: A) -> Retry<I, A> {
+        fn retry_always<A: Action>(_error: &A::Error) -> bool { true };
+        Retry {
+            retry_if: RetryIf::with_condition(handle, strategy, action, retry_always::<A>)
```
Comment (srijs) [inline src/future.rs]: Have you tried making this `...(handle, strategy, action, |_| true as fn(&A::Error) -> bool)` here?

That might help getting rid of `retry_always`.

File: src/future.rs [hunk 2]
```diff
@@ -75,26 +76,57 @@ enum RetryFuturePoll<A> where A: Action {
...
-    handle: Handle
+    handle: Handle,
+    condition: C
 }
 
-impl<I, A> Retry<I, A> where I: Iterator<Item=Duration>, A: Action {
-    pub fn spawn<T: IntoIterator<IntoIter=I, Item=Duration>>(handle: Handle, strategy: T, mut action: A) -> Retry<I, A> {
-        Retry {
+impl<I, A, C> RetryIf<I, A, C> where I: Iterator<Item=Duration>, A: Action, C: Condition<A::Error> {
+    pub fn with_condition<T: IntoIterator<IntoIter=I, Item=Duration>>(
```
Comment (srijs) [inline src/future.rs]: As you mentioned earlier, for consistency's sake this could be just `spawn` now that it lives on another struct.

I've also thought about deprecating the constructors, and just exposing a `retry` function, but that's for another day...

File: src/future.rs (hunk 1 above)
Comment (df5602) [reply src/future.rs]: Thanks for your suggestion. That works and looks much better...

**Comment by df5602** [comment]
I've renamed the constructor and got rid of `retry_always` as you suggested.

Do you want me to rebase?

**Comment by srijs** [comment]
Awesome, thanks! No worries about rebasing, I'll squash your commits when merging.

**Comment by srijs** [comment]
Published your changes as version 0.1.1. Thanks again!

**Comment by df5602** [comment]
Very cool! Thank you for guiding me through this PR 😄
